hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/file/impl/LocalFileSourceFactory.java

```diff
@@ -0,0 +1,231 @@
+import com.hazelcast.jet.pipeline.file.RawBytesFileFormat;
+import com.hazelcast.jet.pipeline.file.TextFileFormat;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import javax.annotation.Nonnull;
+import java.io.BufferedReader;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.Charset;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.stream.Stream;
+
+import static com.hazelcast.jet.datamodel.Tuple2.tuple2;
+import static com.hazelcast.jet.impl.util.Util.uncheckCall;
+import static com.hazelcast.jet.impl.util.Util.uncheckRun;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Implementation of FileSourceFactory for the local filesystem.
+ */
+public class LocalFileSourceFactory implements FileSourceFactory {
+
+    private static Map<String, ReadFileFnProvider> readFileFnProviders;
+
+    static {
+        Map<String, ReadFileFnProvider> mapFns = new HashMap<>();
+
+        addMapFnProvider(mapFns, new JsonReadFileFnProvider());
+        addMapFnProvider(mapFns, new LinesReadFileFnProvider());
+        addMapFnProvider(mapFns, new ParquetReadFileFnProvider());
+        addMapFnProvider(mapFns, new RawBytesReadFileFnProvider());
+        addMapFnProvider(mapFns, new TextReadFileFnProvider());
+
+        ServiceLoader<ReadFileFnProvider> loader = ServiceLoader.load(ReadFileFnProvider.class);
+        for (ReadFileFnProvider readFileFnProvider : loader) {
+            addMapFnProvider(mapFns, readFileFnProvider);
+        }
+
+        LocalFileSourceFactory.readFileFnProviders = Collections.unmodifiableMap(mapFns);
+    }
+
+    private static void addMapFnProvider(Map<String, ReadFileFnProvider> mapFns, ReadFileFnProvider provider) {
+        mapFns.put(provider.format(), provider);
+    }
+
+    @Nonnull @Override
+    public <T> BatchSource<T> create(@Nonnull FileSourceBuilder<T> builder) {
+        Tuple2<String, String> dirAndGlob = deriveDirectoryAndGlobFromPath(builder.path());
+        assert dirAndGlob.f0() != null && dirAndGlob.f1() != null;
+
+        FileFormat<T> format = requireNonNull(builder.format());
+        ReadFileFnProvider readFileFnProvider = readFileFnProviders.get(format.format());
+        FunctionEx<Path, Stream<T>> mapFn = readFileFnProvider.createReadFileFn(format);
+        return Sources.filesBuilder(dirAndGlob.f0())
+                      .glob(dirAndGlob.f1())
+                      .sharedFileSystem(builder.isSharedFileSystem())
+                      .build(mapFn);
+    }
+
+    private Tuple2<String, String> deriveDirectoryAndGlobFromPath(String path) {
+        Path p = Paths.get(path);
+
+        String directory;
+        String glob = "*";
+        if (isDirectory(path)) {
```

Comment:
I think the logic to find if the path is a directory or a file is wrong.  the comment says:
```
        // We can't ask the filesystem because this code runs on the client, which
        // is likely a different machine than the cluster members. So this is a
        // best guess, we assume that directories end with '/'.
```

Let's say I want to read the files in the directory `/user/home/tmp` and I didn't put a `/` to the end.
The source tries to read the file `tmp` in the parent directory `/usr/home`. Since there is no such file in the directory (uses file-name as glob), the source just completes.

Reply:
I see 2 options:
- we either keep the directory and glob parameters instead of single path parameters, for hadoop we would just concatenate these
- we move this piece of code to processor supplier, I have tried to have a go at it, see the latest commit.